include/deal.II/lac/trilinos_vector.h

```diff
@@ -513,85 +599,1562 @@ namespace TrilinosWrappers
+       */
+      DeclException1 (ExcTrilinosError,
+                      int,
+                      << "An error with error number " << arg1
+                      << " occurred while calling a Trilinos function");
+
+      /**
+       * Exception
+       */
+      DeclException4 (ExcAccessToNonLocalElement,
+                      size_type, size_type, size_type, size_type,
+                      << "You tried to access element " << arg1
+                      << " of a distributed vector, but this element is not stored "
+                      << "on the current processor. Note: There are "
+                      << arg2 << " elements stored "
+                      << "on the current processor from within the range "
+                      << arg3 << " through " << arg4
+                      << " but Trilinos vectors need not store contiguous "
+                      << "ranges on each processor, and not every element in "
+                      << "this range may in fact be stored locally.");
+
+    private:
+      /**
+       * Trilinos doesn't allow to mix additions to matrix entries and
+       * overwriting them (to make synchronisation of parallel computations
+       * simpler). The way we do it is to, for each access operation, store
+       * whether it is an insertion or an addition. If the previous one was of
+       * different type, then we first have to flush the Trilinos buffers;
+       * otherwise, we can simply go on.  Luckily, Trilinos has an object for
+       * this which does already all the parallel communications in such a case,
+       * so we simply use their model, which stores whether the last operation
+       * was an addition or an insertion.
+       */
+      Epetra_CombineMode last_action;
+
+      /**
+       * A boolean variable to hold information on whether the vector is
+       * compressed or not.
+       */
+      bool compressed;
+
+      /**
+       * Whether this vector has ghost elements. This is true on all processors
+       * even if only one of them has any ghost elements.
+       */
+      bool has_ghosts;
+
+      /**
+       * Pointer to the actual Epetra vector object. This may represent a vector
+       * that is in fact distributed among multiple processors. The object
+       * requires an existing Epetra_Map for storing data when setting it up.
+       */
+      std::shared_ptr<Epetra_FEVector> vector;
+
+      /**
+       * A vector object in Trilinos to be used for collecting the non-local
+       * elements if the vector was constructed with an additional IndexSet
+       * describing ghost elements.
+       */
+      std::shared_ptr<Epetra_MultiVector> nonlocal_vector;
+
+      /**
+       * An IndexSet storing the indices this vector owns exclusively.
+       */
+      IndexSet owned_elements;
+
+      /**
+       * Make the reference class a friend.
+       */
+      friend class internal::VectorReference;
+      friend class Vector;
+      friend class MPI::Vector;
```

Comment:
We should remove the last two friendship declarations: `TrilinosWrappers::Vector` no longer exists and classes are always friends with themselves.